### DISCUSSION TITLE
How to declare the validated fields as Final ?

### DISCUSSION THREAD

**Discussion by vtgn** [pydantic/pydantic#10860] [Question] [answered]
Hi!

I know there is the `frozen` setting to generate read-only fields, but the fields in the Pydantic model can't be defined as Final because Pydantic ignore them. So the validated instances of the model have read-only fields, but the typing linters don't know and accept to reassign the fields, which doesn't work during execution of course.
Other information: I would like to use one class only, to validate and store correctly the fields the way I want.

### How to declare the validated fields as Final?
Why Pydantic ignore the fields in the model if they are typed as Final? Final is just a static information and have no effect during execution, allowing to a Final variable to be reassigned.
If Pydantic accepted Final fields, we could declare real frozen objects having read-only fields statically and dynamically.

### Other question: Why Pydantic refuse to validate private attributes?
I ask this question because maybe there is a good reason for that, but I just want to be sure this reason exists.
Indeed, I'm disappointed to not be able to use this feature because it could be useful in some cases.
I tried this solution to avoid the non final public declared fields by using private ones and defining public getter on them.

Thanks.

Regard.

**Comment by Viicos** [comment, marked as answer]
See my comment [here](https://github.com/pydantic/pydantic/issues/10474#issuecomment-2478666651) which is closely related. tl;dr, `Final` is not treated as it should by Pydantic currently.

**Reply by vtgn** [reply]
Thanks for this information. Shame this is not possible yet. :/
For now, I've just done 2 classes: one for the Pydantic model, the other for declaring the attributes as Final and setting their values from the constructor by passing the dictionary of validated values by the Pydantic model. I've had to ignore the typing errors from the typing linters (Pylance and mypy) on the declared Final attributes not set immediately by adding the `# type: ignore` comment on each attribute's declaration line.
